### PR TITLE
[WIP] Support weight decay in SGD optimizer when gradient is sparse

### PR DESCRIPTION
**Pull request by ezyang**
Towards fixing #1285.

This is a WIP commit to get the ball rolling on code review
(I am sure I have done great violence to the various coding
standards of your project.)

Things to be done:

    - Continue adding sparse support for other parameters
      and optimizers

    - Add some more tests, including a unit test ensuring
      that a single step is what we expect

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

CC @adamlerer @soumith

### CODE REVIEW

**Comment by apaszke** [comment]
I wonder if it wouldn't be simpler and less magical to offer that deferred sparse weight decay as an alternative kind of an optimizer. The user would create the regular one with weight_decay=0 and then another one `wd = optim.DeferredWD(model.parameters())`. It seems more explicit, otherwise it might be easy to forget about `flush()` because right now the user doesn't really need to be aware which gradients are sparse.

**Comment by ezyang** [comment]
That's OK by me. Would you suggest just erroring (with a suggestion) if they try to use weight decay while optimizing something sparse? (@adamlerer, what do you think?)

**Comment by graytowne** [comment]
Hi @ezyang,
Can I know if the weight decay for sparse gradient is supported currently? 
If not, is there any other alternated ways to perform it?

I'm working on recommendation, and the weight decay for sparse embeddings is crucial as models are easily get overfitted.

**Comment by ezyang** [comment]
No, it's not been merged yet. But it shouldn't be too hard to adapt this patch set to get it going for what you need. Give it a shot maybe?

**Comment by graytowne** [comment]
OK, I will take a try, thanks so much! One simple question is: [Here](https://github.com/pytorch/pytorch/pull/1305/files#diff-d237dd23d765aba7cb30b4c21a5e1a86R104) why it needs to call flush() every time? I think the dense tensor doesn't need this.
